### PR TITLE
unit: skip battery check when systemd.skip-battery-check specified on boot

### LINKED ISSUE #28366: proc-cmdline: calling getopt_long() after proc_cmdline_filter_pid1_args() causes use-after-free
### systemd version the issue has been seen with

HEAD (v254-rc1)

### Used distribution

_No response_

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

With the patch below,
```
$ sudo valgrind --leak-check=full ./systemd-battery-check --help
==92765== Memcheck, a memory error detector
==92765== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==92765== Using Valgrind-3.21.0 and LibVEX; rerun with -h for copyright info
==92765== Command: ./systemd-battery-check --help
==92765== 
before: optind=1, opterr=1, optopt=63, optarg=n/a
after: optind=1, opterr=1, optopt=63, optarg=n/a
==92765== Invalid read of size 1
==92765==    at 0x4E3B6B0: _getopt_internal_r (in /usr/lib64/libc.so.6)
==92765==    by 0x4E3BC5F: _getopt_internal (in /usr/lib64/libc.so.6)
==92765==    by 0x4E3BCE5: getopt_long (in /usr/lib64/libc.so.6)
==92765==    by 0x402E25: parse_argv (battery-check.c:106)
==92765==    by 0x403123: run (battery-check.c:149)
==92765==    by 0x403601: main (battery-check.c:194)
==92765==  Address 0x5a47dc2 is 2 bytes inside a block of size 3 free'd
==92765==    at 0x48442AC: free (vg_replace_malloc.c:974)
==92765==    by 0x4B0A892: strv_free (strv.c:79)
==92765==    by 0x40281C: strv_freep (strv.h:28)
==92765==    by 0x403100: run (battery-check.c:139)
==92765==    by 0x403601: main (battery-check.c:194)
==92765==  Block was alloc'd at
==92765==    at 0x484182F: malloc (vg_replace_malloc.c:431)
==92765==    by 0x4DE7FEE: strdup (in /usr/lib64/libc.so.6)
==92765==    by 0x4B0ABB2: strv_new_ap (strv.c:153)
==92765==    by 0x4B0AD4E: strv_new_internal (strv.c:170)
==92765==    by 0x402F70: run (battery-check.c:139)
==92765==    by 0x403601: main (battery-check.c:194)
==92765== 
systemd-battery-check

Check battery level to see whether there's enough charge.

   -h --help            Show this help
      --version         Show package version

See the systemd-battery-check(8) man page for details.
==92765== 
==92765== HEAP SUMMARY:
==92765==     in use at exit: 0 bytes in 0 blocks
==92765==   total heap usage: 58 allocs, 58 frees, 15,082 bytes allocated
==92765== 
==92765== All heap blocks were freed -- no leaks are possible
==92765== 
==92765== For lists of detected and suppressed errors, rerun with: -s
==92765== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```
As `proc_cmdline_filter_pid1_args()` is called in `proc_cmdline_parse()` when running in container, so, parsing arguments will fail, see e.g. https://github.com/systemd/systemd/actions/runs/5526400032/jobs/10081093095?pr=28355

### Steps to reproduce the problem

```diff
diff --git a/src/battery-check/battery-check.c b/src/battery-check/battery-check.c
index ca13251ae0..d20d55d2a6 100644
--- a/src/battery-check/battery-check.c
+++ b/src/battery-check/battery-check.c
@@ -15,7 +15,9 @@
 #include "log.h"
 #include "main-func.h"
 #include "pretty-print.h"
+#include "proc-cmdline.h"
 #include "socket-util.h"
+#include "strv.h"
 #include "terminal-util.h"
 #include "time-util.h"
 
@@ -132,6 +134,18 @@ static int run(int argc, char *argv[]) {
 
         log_setup();
 
+        {
+                _cleanup_strv_free_ char **filtered = NULL,
+                        **args = strv_new("-h", "aaa", "--version", "-n", "unit", "aaaa", "foo", "-h");
+
+                log_info("before: optind=%i, opterr=%i, optopt=%i, optarg=%s", optind, opterr, optopt, strna(optarg));
+
+                assert_se(args);
+                assert_se(proc_cmdline_filter_pid1_args(args, &filtered) >= 0);
+
+                log_info("after: optind=%i, opterr=%i, optopt=%i, optarg=%s", optind, opterr, optopt, strna(optarg));
+        }
+
         r = parse_argv(argc, argv);
         if (r <= 0)
                 return r;
```

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
This also fixes #28366.

### CODE REVIEW

**Review by YHNdnzj — APPROVED** [review]
Maybe also add this to the NEWS?

**Comment by DaanDeMeyer** [comment]
Shouldn't this just be systemd.battery-check= which takes a boolean?

**Comment by yuwata** [comment]
OK, now the kernel command line option is parsed by the command, instead of by `ConditionKernelCommandLine=`, and the option takes boolean value. PTAL.

**Comment by poettering** [comment]
lgtm, but I think it would make sense to also add something like this to the unit file, to bypass it in the common case.

```
ConditionKernelCommandLine=!systemd.battery-check=0
```

of course, there are multiple ways to encode "off" for this, I'd just write down the most common one here (which I think is "0") and make sure that the docs suggest this syntax, and not the others.

This condition is an optimization only of course, which is why I think it is fine to just list the most common way to write this...

File: man/systemd-battery-check.service.xml [hunk 1]
```diff
@@ -62,6 +62,26 @@
...
+    <para>The following variables are understood:</para>
+
+    <variablelist class='kernel-commandline-options'>
+      <varlistentry>
+        <term><varname>systemd.battery-check=<replaceable>BOOL</replaceable></varname></term>
+
+        <listitem>
+          <para>Takes a boolean. If specified with false, <command>systemd-battery-check</command> command
+          will return immediately with exit status 0 without checking battery capacity and AC power
+          existence, and the service <filename>systemd-battery-check.service</filename> will succeeds. This
```
Comment (bluca) [inline man/systemd-battery-check.service.xml]: "will succeed"

**Review by bluca — APPROVED** [review]

File: man/systemd-battery-check.service.xml (hunk 1 above)
Comment (bluca) [reply man/systemd-battery-check.service.xml]: given it's just a typo, I fixed it and merged

**Pull request merged**